### PR TITLE
Add deterministic live supervisor

### PR DESCRIPTION
**Pull request by phroi**
## Why

Live testnet runs need a deterministic supervisor that can exercise bounded bot/tester cycles, classify known outcomes, and leave secret-safe artifacts for operator review without embedding repair or LLM behavior in the funded actor path.

## Changes

- Add `@ickb/supervisor` with bounded scenario planning, outcome classification, coverage ledgers, incident bundles, and dry-run fixtures.
- Add an external summary-only supervisor loop for repeated bounded invocations.
- Wire `pnpm live:supervisor` and `pnpm live:supervisor:loop`, include supervisor tests in Vitest wiring, ignore `logs/live-supervisor/`, and document the deterministic external-repair boundary.

### CODE REVIEW

File: apps/supervisor/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,2047 @@
...
+}
+
+function parsePositiveInteger(value: string, flag: string): number {
+  if (!/^[1-9][0-9]*$/u.test(value)) {
+    throw new Error(`Invalid ${flag}: expected a positive integer`);
+  }
+  const parsed = Number(value);
+  if (!Number.isSafeInteger(parsed)) {
+    throw new Error(`Invalid ${flag}: expected a safe integer`);
+  }
```
[the remark this answers is not in the document]
Comment (phroi) [reply apps/supervisor/src/index.ts]: Updated the supervisor integer parser to validate digit strings with BigInt before converting to Number. Values above Number.MAX_SAFE_INTEGER now fail without rounding through Number first, and the supervisor tests cover the boundary.

File: apps/supervisor/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,2047 @@
...
+  return parsed;
```
[the remark this answers is not in the document]
Comment (phroi) [reply apps/supervisor/src/index.ts]: This is covered by the same parser change: the supervisor now compares the parsed BigInt against Number.MAX_SAFE_INTEGER before returning a Number, with tests for max-safe acceptance and unsafe rejection.

File: scripts/ickb-supervisor-loop.mjs [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+}
+
+function parseNonNegativeInteger(value, flag) {
+  if (!/^(0|[1-9][0-9]*)$/u.test(value)) {
+    throw new Error(`Invalid ${flag}: expected a non-negative integer`);
+  }
+  const parsed = Number(value);
+  if (!Number.isSafeInteger(parsed)) {
+    throw new Error(`Invalid ${flag}: expected a safe integer`);
+  }
```
[the remark this answers is not in the document]
Comment (phroi) [reply scripts/ickb-supervisor-loop.mjs]: Updated the loop non-negative parser the same way: regex validation first, BigInt safe-range check second, then Number conversion only after the value is known safe.

File: scripts/ickb-supervisor-loop.mjs [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+
+async function readSummary(path, dependencies) {
+  const readFileFn = dependencies.readFile ?? readFile;
+  const text = await readFileFn(path, "utf8");
+  let parsed;
+  try {
+    parsed = JSON.parse(text);
+  } catch {
+    throw new Error("summary.json invalid JSON");
+  }
```
[the remark this answers is not in the document]
Comment (phroi) [reply scripts/ickb-supervisor-loop.mjs]: Updated readSummary to rethrow invalid JSON as a safe summary.json error while preserving the original parse error as the cause. The user-facing loop output still does not include the raw summary contents.

File: scripts/ickb-supervisor-loop.mjs [hunk 5]
```diff
@@ -0,0 +1,443 @@
...
+}
+
+function parsePositiveInteger(value, flag) {
+  if (!/^[1-9][0-9]*$/u.test(value)) {
+    throw new Error(`Invalid ${flag}: expected a positive integer`);
+  }
+  const parsed = Number(value);
+  if (!Number.isSafeInteger(parsed)) {
+    throw new Error(`Invalid ${flag}: expected a safe integer`);
+  }
```
[the remark this answers is not in the document]
Comment (phroi) [reply scripts/ickb-supervisor-loop.mjs]: Updated the loop positive-integer parser to validate with BigInt before converting to Number. The loop tests now reject values above Number.MAX_SAFE_INTEGER without relying on rounded Number conversion.

File: scripts/ickb-supervisor-loop.mjs [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+function parseNonNegativeInteger(value, flag) {
+  if (!/^(0|[1-9][0-9]*)$/u.test(value)) {
+    throw new Error(`Invalid ${flag}: expected a non-negative integer`);
+  }
+  const parsed = Number(value);
+  if (!Number.isSafeInteger(parsed)) {
+    throw new Error(`Invalid ${flag}: expected a safe integer`);
+  }
+  return parsed;
+}
```
[the remark this answers is not in the document]
Comment (phroi) [reply scripts/ickb-supervisor-loop.mjs]: Applied to both loop integer helpers. They now validate with BigInt against Number.MAX_SAFE_INTEGER before returning a Number, and the loop tests cover unsafe large values.

File: scripts/ickb-supervisor-loop.mjs [hunk 7]
```diff
@@ -0,0 +1,443 @@
...
+
+function parsePositiveInteger(value, flag) {
+  if (!/^[1-9][0-9]*$/u.test(value)) {
+    throw new Error(`Invalid ${flag}: expected a positive integer`);
+  }
+  const parsed = Number(value);
+  if (!Number.isSafeInteger(parsed)) {
+    throw new Error(`Invalid ${flag}: expected a safe integer`);
+  }
+  return parsed;
```
[the remark this answers is not in the document]
Comment (phroi) [reply scripts/ickb-supervisor-loop.mjs]: The loop positive-integer parser now uses BigInt for safe-range validation before Number conversion. This closes the precision-loss case for very large digit strings.

File: scripts/ickb-supervisor-loop.mjs [hunk 8]
```diff
@@ -0,0 +1,443 @@
...
+    return { action: "stop", reason: "max_runs", newOutcomes, stableCount: nextStableCount, exitCode: 0 };
+  }
+  return { action: "continue", reason: "continue", newOutcomes, stableCount: nextStableCount, exitCode: 0 };
+}
+
+export function summarySignature(summary) {
+  return JSON.stringify({
+    stopped: typeof summary.stopped === "string" ? summary.stopped : "unknown",
+    aggregateCounts: sortedEntries(countRecord(summary.aggregateCounts)),
+    skipReasons: stringArray(summary.skipReasons),
```
[the remark this answers is not in the document]
Comment (phroi) [reply scripts/ickb-supervisor-loop.mjs]: summarySignature now sorts skipReasons before serialization, so equivalent skip-reason sets produce the same stable-no-progress signature. Added a loop test for differing skip-reason order.

File: apps/supervisor/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,2047 @@
...
+  result: CommandResult,
+  evidence: ParsedEvidence,
+  base: Omit<Classification, "outcome" | "terminal" | "reason">,
+): Classification {
+  if (result.status !== 0) {
+    return {
+      ...base,
+      outcome: result.stderr.includes("Invalid") && result.stderr.includes("chain identity") ? "wrong_chain" : "nonzero_exit",
+      terminal: true,
+      reason: boundedText(safeArtifactText(result.stderr), 240) || "preflight command exited nonzero",
```
[the remark this answers is not in the document]
Comment (phroi) [reply apps/supervisor/src/index.ts:1104]: Spawn failures now classify before generic evidence handling, so preflight spawn errors produce a terminal nonzero_exit with an actor-specific reason instead of a generic nonzero message.

File: apps/supervisor/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,2047 @@
...
+    };
+  }
+
+  if (result.status !== 0) {
+    return {
+      ...base,
+      outcome: "nonzero_exit",
+      terminal: true,
+      reason: `bot exited with status ${String(result.status)}`,
+      publicState,
```
[the remark this answers is not in the document]
Comment (phroi) [reply apps/supervisor/src/index.ts:1168]: Bot spawn failures now take the same explicit path: the classifier returns terminal nonzero_exit with the captured spawnError in the reason before falling back to generic exit-status handling.

File: apps/supervisor/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,2047 @@
...
+      return {
+        ...base,
+        outcome: "tester_deterministic_pre_broadcast_error",
+        terminal: true,
+        reason: "tester recorded an error before a committed transaction was proven",
+      };
+    }
+  }
+  if (result.status !== 0) {
+    return { ...base, outcome: "nonzero_exit", terminal: true, reason: `tester exited with status ${String(result.status)}` };
```
[the remark this answers is not in the document]
Comment (phroi) [reply apps/supervisor/src/index.ts:1241]: Tester spawn failures are now handled by the shared spawn-error branch as well, so they report terminal nonzero_exit with the captured spawnError rather than a null-status generic exit reason.

File: apps/supervisor/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,2047 @@
...
+  }
+  const parsed = Number(value);
+  if (!Number.isSafeInteger(parsed)) {
+    throw new Error(`Invalid ${flag}: expected a safe integer`);
+  }
+  return parsed;
+}
+
+function parseTesterFeeValue(value: string, flag: string): string {
+  if (!/^(?:0|[1-9][0-9]*)$/u.test(value)) {
```
[the remark this answers is not in the document]
Comment (phroi) [reply apps/supervisor/src/index.ts:1905]: Changed the built-runtime preflight to check the canonical local CCC build outputs under forks/ccc/repo/packages/core and forks/ccc/repo/packages/udt, avoiding package-local node_modules layout assumptions.

File: apps/supervisor/README.md [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+`--tester-scenario` is passed to the tester as `TESTER_SCENARIO`. When it is left as `auto`, `--target-outcome tester_conversion_created` steers the tester to `sdk-conversion`, the SDK conversion-builder selector. Use `ickb-to-ckb-limit-order` for iCKB withdrawal-through-LO coverage; `all-ickb-limit-order` remains a compatibility alias. Use `sdk-conversion` when the intended behavior is the SDK conversion builder, including direct conversions that do not create limit orders, `multi-order-limit-orders` for any funded two-order raw limit-order type, `two-ckb-to-ickb-limit-orders`, `two-ickb-to-ckb-limit-orders`, or `mixed-direction-limit-orders` for a specific two-order transaction, and `extra-large-limit-order` to stress non-interface users placing large raw limit orders. An explicit `--tester-scenario` overrides target-outcome steering.
```
[the remark this answers is not in the document]
Comment (phroi) [reply apps/supervisor/README.md]: The current README/source already list both directions separately: `two-ckb-to-ickb-limit-orders` and `two-ickb-to-ckb-limit-orders`. I checked the live tree and there is no duplicate `two-ckb-to-ickb-limit-orders` entry to correct.

File: apps/supervisor/README.md [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+
+The supervisor treats public testnet iCKB deposits, receipts, and orders as observable scenario surface. It does not count public state as bot/tester-owned inventory or as permission to mutate unrelated cells.
+
+## Scenario Coverage
+
+The default planner prefers under-covered safe outcomes over repeating the same known-good path. Supported selectors are:
+
+```bash
+--scenario auto|standard-cycle|tester-only|bot-only|tester-fresh-skip-two-pass
+--tester-scenario auto|random-order|sdk-conversion|extra-large-limit-order|multi-order-limit-orders|two-ckb-to-ickb-limit-orders|all-ckb-limit-order|all-ickb-limit-order|ickb-to-ckb-limit-order|two-ickb-to-ckb-limit-orders|mixed-direction-limit-orders|dust-ckb-conversion|dust-ickb-conversion
```
[the remark this answers is not in the document]
Comment (phroi) [reply apps/supervisor/README.md]: The README is listing two accepted selector names, not a duplicate value. `all-ickb-limit-order` is intentionally documented on this line as a compatibility alias for `ickb-to-ckb-limit-order`, and both are accepted by the tester/supervisor scenario parsers.

[the remark this answers is not in the document]
Comment (phroi) [reply apps/supervisor/README.md]: No change here: this is the same documented compatibility alias as the earlier thread. `all-ickb-limit-order` and `ickb-to-ckb-limit-order` are both accepted selector names, and the paragraph below the list states that `all-ickb-limit-order` remains a compatibility alias.

File: scripts/ickb-supervisor-loop.mjs [hunk 15]
```diff
@@ -0,0 +1,443 @@
...
+  "bot_match_committed",
+  "bot_match_plus_deposit_committed",
+  "bot_receipt_completion_committed",
+  "bot_deposit_only_committed",
+  "bot_withdrawal_request_committed",
+  "bot_withdrawal_completion_committed",
+  "confirmation_timeout",
+  "terminal_chain_rejection",
+  "post_broadcast_unresolved",
+]);
```
[the remark this answers is not in the document]
Comment (phroi) [reply scripts/ickb-supervisor-loop.mjs]: Fixed. The supervisor now exports `TX_CREATING_OUTCOMES`, and the loop imports it from `apps/supervisor/dist/index.js`, so the supervisor summary and loop stop condition share one source of truth.

File: apps/tester/src/index.ts [hunk 16]
```diff
@@ -209,24 +209,19 @@ async function main(): Promise<void> {
...
-            reserve: formatCkb(CKB_RESERVE),
-            postTxCkbBalance: formatCkb(postTxCkbBalance),
-          };
-          if (logTerminalIteration(executionLog, startTime, ++completedIterations, maxIterations)) {
-            return;
-          }
-          continue;
+      const postTxCkbBalance = postTransactionPlainCkbBalance(tx, state, runtime.accountLocks);
+      const reserveSkip = testerReserveSkip(postTxCkbBalance);
+      if (reserveSkip !== undefined) {
```
[the remark this answers is not in the document]
Comment (phroi) [reply apps/tester/src/index.ts]: Fixed. The tester reserve gate is again conditional on planned CKB-consuming orders (`ckb-to-ickb`), so iCKB-to-CKB replenishment paths are not blocked just because the final plain CKB balance is still below the reserve.

File: apps/supervisor/src/index.ts [hunk 17]
```diff
@@ -0,0 +1,2066 @@
...
+    let timedOut = false;
+    let settled = false;
+    let killTimeout: NodeJS.Timeout | undefined;
+    const timeout = setTimeout(() => {
+      timedOut = true;
+      signalChild(child, "SIGTERM", dependencies);
+      killTimeout = setTimeout(() => {
+        signalChild(child, "SIGKILL", dependencies);
+      }, dependencies.commandKillGraceMs ?? DEFAULT_COMMAND_KILL_GRACE_MS);
+    }, spec.timeoutMs);
```
[the remark this answers is not in the document]
Comment (phroi) [reply apps/supervisor/src/index.ts]: Fixed. Actor command timeout and kill-grace delays are capped before calling `setTimeout`, and the supervisor tests now cover an oversized timeout/grace value.

File: apps/tester/src/index.ts [hunk 18]
```diff
@@ -209,24 +209,19 @@ async function main(): Promise<void> {
...
+          throw new TesterTerminalError(
+            `Not enough CKB to preserve tester reserve after the tx: expected ${formatCkb(CKB_RESERVE)} CKB, got ${formatCkb(postTxCkbBalance)} CKB`,
+          );
+        }
+        executionLog.skip = reserveSkip;
+        if (logTerminalIteration(executionLog, startTime, ++completedIterations, maxIterations)) {
+          return;
         }
+        continue;
       }
```
[the remark this answers is not in the document]
Comment (phroi) [reply apps/tester/src/index.ts]: Fixed. The post-transaction reserve skip now runs only when the planned tester orders include a CKB-to-iCKB order, preserving the replenishment behavior for iCKB-to-CKB transactions.

File: apps/supervisor/src/index.ts [hunk 19]
```diff
@@ -0,0 +1,2066 @@
...
+        ...result,
+      });
+    };
+
+    child.stdout.on("data", (chunk: Buffer) => {
+      appendBoundedOutput(stdoutCapture, chunk, maxOutputBytes);
+    });
+    child.stderr.on("data", (chunk: Buffer) => {
+      appendBoundedOutput(stderrCapture, chunk, maxOutputBytes);
+    });
```
[the remark this answers is not in the document]
Comment (phroi) [reply apps/supervisor/src/index.ts:1049]: No change needed for the current invariant. This `spawn` call owns its `stdio` shape and always requests `pipe` for stdout and stderr; adding optional chaining is rejected by the repo lint rule as an unnecessary condition.

File: apps/supervisor/src/index.ts [hunk 20]
```diff
@@ -0,0 +1,2066 @@
...
+
+function boundedText(text: string, limit: number): string {
+  if (text.length <= limit) {
+    return text;
+  }
+  return `${text.slice(0, limit)}\n<truncated ${String(text.length - limit)} bytes>`;
+}
+
+function errorMessage(error: unknown): string {
+  return error instanceof Error ? error.message : typeof error === "string" ? error : "Unknown error";
```
[the remark this answers is not in the document]
Comment (phroi) [reply apps/supervisor/src/index.ts:2026]: No code change here. `errorMessage` already handles real `Error` instances and string throws without unsafe casts; broadening formatting for arbitrary thrown records is not needed for the supervisor safety boundary.

File: apps/supervisor/src/index.ts [hunk 21]
```diff
@@ -0,0 +1,2066 @@
...
+}
+
+function isNotFoundError(error: unknown): boolean {
+  return isRecord(error) && error["code"] === "ENOENT";
```
[the remark this answers is not in the document]
Comment (phroi) [reply apps/supervisor/src/index.ts:2030]: No code change here. `isNotFoundError` first narrows with `isRecord`, and the indexed `code` check is type-safe and behaviorally equivalent for this local ENOENT guard.

File: apps/supervisor/src/index.ts [hunk 22]
```diff
@@ -0,0 +1,2066 @@
...
+  "command_timeout",
+  "nonzero_exit",
+  "unmet_coverage_goal",
+  "unknown",
+] as const;
+
+export type OutcomeKind = typeof OUTCOME_KINDS[number];
+export type Actor = "bot" | "tester";
+export type ScenarioName = "auto" | "standard-cycle" | "tester-only" | "bot-only" | "tester-fresh-skip-two-pass";
+export type TesterScenario = "auto" | "random-order" | "sdk-conversion" | "extra-large-limit-order" | "multi-order-limit-orders" | "two-ckb-to-ickb-limit-orders" | "all-ckb-limit-order" | "all-ickb-limit-order" | "ickb-to-ckb-limit-order" | "two-ickb-to-ckb-limit-orders" | "mixed-direction-limit-orders" | "dust-ckb-conversion" | "dust-ickb-conversion";
```
[the remark this answers is not in the document]
Comment (phroi) [reply apps/supervisor/src/index.ts]: No change: `all-ickb-limit-order` is an intentional compatibility alias, not a duplicate of `ickb-to-ckb-limit-order`. The README documents that alias, and both the supervisor and tester parsers accept both selector names.

File: apps/supervisor/README.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (phroi) [reply apps/supervisor/README.md]: No change: this is the same compatibility-alias issue already checked in the current tree. `all-ickb-limit-order` is intentionally documented as an accepted alias for `ickb-to-ckb-limit-order`; removing it would drop an existing selector name rather than remove a duplicate.

File: apps/supervisor/src/index.ts [hunk 23]
```diff
@@ -0,0 +1,2073 @@
...
```
[the remark this answers is not in the document]
Comment (phroi) [reply apps/supervisor/src/index.ts]: No change for the same reason: the type, usage text, and parser list two accepted selector names. `all-ickb-limit-order` is the compatibility alias and `ickb-to-ckb-limit-order` is the preferred selector, so this is intentional current behavior.

**Comment by phroi** [comment]
LGTM

Phroi %261
